src/metabase/driver/mysql.clj

```diff
@@ -617,7 +619,30 @@
     ::upload/float                    [:double]
     ::upload/boolean                  [:boolean]
     ::upload/date                     [:date]
-    ::upload/datetime                 [:timestamp]))
+    ::upload/datetime                 [:timestamp]
+    ::upload/zoned-datetime           [:timestamp]))
+
+(defn- offset-date-time->local-date-time
+  "Remove the offset from a string datetime, returning a LocalDateTime in whatever timezone the `database` is configured
+  to use. This is necessary since MySQL doesn't support timezone offsets and so we need to calculate one by hand. This
+  is used for the upload parser below, which was looking messy with error-handling so it made sense to extract the logic."
+  [s database]
+  (let [offset-time (t/offset-date-time (str/trim s))
+        zone-id     (t/zone-id (qp.timezone/results-timezone-id database))]
+    (t/local-date-time offset-time zone-id )))
+
+(defmethod driver/upload-type->parser [:mysql :metabase.upload/zoned-datetime]
```

Comment:
I don't think you need this. MySQL does support timezone offsets on insertion. It just doesn't store the time zone it was inserted with. A string like '2023-10-19 10:00:00+00:00' should be able to be inserted into MySQL directly.

I just tested this and it works as I expected.
```
CREATE TABLE timestamps (
    event_timestamp TIMESTAMP NOT NULL
);

INSERT INTO timestamps (event_timestamp) VALUES ('2023-10-19 10:00:00+00:00'), ('2023-10-19 10:00:00+01:00');

SELECT * from timestamps;
```
The result is:
```
+---------------------+
| event_timestamp     |
+---------------------+
| 2023-10-19 13:00:00 |
| 2023-10-19 12:00:00 |
+---------------------+
```
Because my system timezone offset is UTC+3 right now

Reply:
I was led to believe that as well from the docs, but this test was failing:


```clojure
           (csv-file-with ["offset_datetime"
                           "2022-01-01T01:00:00-07:00"])))
...
            (is (= "2022-01-01T08:00:00Z"
                   (second (first (rows-for-table table)))))))))))
```


```
Fail in load-from-csv-offset-datetime-test
Upload a CSV file with a datetime column 
...

expected: "2022-01-01T08:00:00Z"

  actual: "2022-01-01T01:00:07Z"	  
    diff: - "2022-01-01T08:00:00Z"	  
          + "2022-01-01T01:00:07Z"	    
```

Treating the offset as a change to the _seconds_ is weird and maybe java-time's fault instead of MySQLs.

That to one side, I believe the offset support was introduced in MySQL 8 ([link](https://github.com/metabase/metabase/issues/34137#issuecomment-1744993754)), so making it explicit the way it is here maintains support for earlier versions. It happens that MySQL < 8 EOLs in about two weeks, so we're allowed to break support, but "write special logic to make extra sure we're handling offsets right and also support old versions" seemed more appealing than trying to figure out what's up with java-time.

Reply:
This isn't an issue with java-time. It's an issue with `metabase.driver.mysql/row->tsv` and how MySQL ingests timestamps from TSVs.

I ran the `load-from-csv-offset-datetime-test` test with this `git diff`:

```diff
diff --git a/src/metabase/driver/mysql.clj b/src/metabase/driver/mysql.clj
index 05f502e502..1431c0188b 100644
--- a/src/metabase/driver/mysql.clj
+++ b/src/metabase/driver/mysql.clj
@@ -631,7 +631,7 @@
         zone-id     (t/zone-id (qp.timezone/results-timezone-id database))]
     (t/local-date-time offset-time zone-id )))
 
-(defmethod driver/upload-type->parser [:mysql ::upload/offset-datetime]
+#_(defmethod driver/upload-type->parser [:mysql ::upload/offset-datetime]
   [_ _]
   (comp
    (fn [s]
@@ -699,7 +699,7 @@
     (let [temp-file (File/createTempFile table-name ".tsv")
           file-path (.getAbsolutePath temp-file)]
       (try
-        (let [tsvs (map (partial row->tsv (count column-names)) values)
+        (let [tsvs (dev/tap>-spy (map (partial row->tsv (count column-names)) (dev/tap>-spy values)))
               sql  (sql/format {::load   [file-path (keyword table-name)]
                                 :columns (map keyword column-names)}
                                :quoted  true
```

The tapped result:
```
1st tap (values):
((#t "2022-01-01T01:00-07:00"))

2nd tap (map (partial row->tsv (count column-names)) (dev/tap>-spy values))
("2022-01-01T01:00-07:00")
```

Then I tested how this value would get converted from MySQL:

With this test CSV:
```
event_timestamp
"2022-01-01T01:00:00-07:00"
"2022-01-01T01:00-07:00"
```

And this SQL:
```
DROP TABLE IF EXISTS my_table;

CREATE TABLE my_table (
    event_timestamp TIMESTAMP
);

LOAD DATA LOCAL INFILE '/Users/callumherries/meta/metabase/test.csv'
INTO TABLE my_table
FIELDS TERMINATED BY ','
ENCLOSED BY '"'
LINES TERMINATED BY '\n'
IGNORE 1 ROWS
(event_timestamp);

SELECT * FROM my_table;
```

The result:
```
+---------------------+
| event_timestamp     |
+---------------------+
| 2022-01-01 10:00:00 |
| 2022-01-01 01:00:07 |
+---------------------+
2 rows in set (0.00 sec)
```

Reply:
Thanks for looking into that so exhaustively.

Especially since [we're off of MySQL < 8 now](https://metaboat.slack.com/archives/C010L1Z4F9S/p1698076249487039) I fixed the TSV generation as indicated, and moved `upload-type->parser` out of driver-land and into `upload.parsing`.